python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+        self,
+        dimension_packer: DimensionPacker,
+        data_id: DataCoordinate,
+        release_id: int = DEFAULT_RELEASE_ID,
+        n_releases: int = DEFAULT_N_RELEASES,
+    ):
+        self._bits = _IdGeneratorBits(dimension_packer, n_releases)
+        self._release_id = release_id
+        self._data_id = data_id.subset(self._bits.packer.dimensions)
+        self._packed_data_id = self._bits.packer.pack(self._data_id)
+
+    @property
+    def data_id(self) -> DataCoordinate:
+        """The data ID that will be embedded in all generated IDs
+        (`DataCoordinate`)."""
+        return self._data_id
+
+    @property
+    def release_id(self) -> int:
+        """The release ID that will embedded in all generated IDs (`int`)."""
+
+    @property
+    def catalog_id(self) -> int:
+        # Docstring inherited.
+        return self._packed_data_id + self._bits.n_data_ids * self._release_id
+
+    def __str__(self) -> str:
+        # Docstring inherited.
+        return str(self.data_id)
+
+    def make_table_id_factory(self) -> IdFactory:
+        # Docstring inherited.
+        return IdFactory.makeSource(self.catalog_id, self._bits.counter_bits)
+
+    def arange(self, *args, **kwargs) -> np.ndarray:
+        # Docstring inherited.
+        lower = super().arange(*args, **kwargs)
+        if np.any(lower >= self._bits.n_counters):
+            arg_terms = [repr(arg) for arg in args] + [
+                f"{k}={v!r}" for k, v in kwargs.items()
+            ]
+            raise ValueError(
+                f"Integer range from numpy.arange({arg_terms}) has values that are not "
+                f"below upper bound of {self._bits.n_counters}."
+            )
+        return lower + self.catalog_id * self._bits.n_counters
+
+
+@dataclasses.dataclass
+class _IdGeneratorBits:
+    """A private helper struct that manages the allocation of bits between the
+    packed data ID, the release ID, and a per-catalog counter.
+    """
+
+    packer: DimensionPacker
+    """Object that maps data IDs to integers
+    (`lsst.daf.butler.DimensionPacker`).
+    """
+
+    n_releases: int = dataclasses.field(default=0)
+    """Number of releases to reserve space for, starting from zero (`int`)."""
+
+    n_data_ids: int = dataclasses.field(init=False)
+    """Number of contiguous packed data IDs to reserve space for, starting
+    from zero (`int`).
+    """
+
+    counter_bits: int = dataclasses.field(init=False)
+    """Number of bits allocated to the per-catalog counter (`int`)."""
+
+    n_counters: int = dataclasses.field(init=False)
+    """Number of contiguous counter values to reserve space for, starting from
+    zero (`int`)."""
+
+    def __post_init__(self) -> None:
+        self.n_data_ids = 1 << self.packer.maxBits
+        upper_bits = (self.n_releases - 1).bit_length() + self.packer.maxBits
+        self.counter_bits = IdFactory.computeReservedFromMaxBits(upper_bits)
+        self.n_counters = 1 << self.counter_bits
```

Comment:
The class docstring should probably note these three attributes, even though this is intended to be private.

Reply:
What do you mean?  I don't think of attribute docs as going in the class docstring at all; the class `Parameters` is for `__init__` arguments, and these are not.

Reply:
Sorry about that: I see them as `init=False` fields now, with docstrings, so that's fine.